consensus/service/src/consensus_service.rs

```diff
@@ -471,8 +482,7 @@ impl<
                     );
 
                     // Keep track that we heard from the sender of this message.
-                    {
-                        let peer_keepalive = peer_keepalive.lock().expect("mutex poisoned");
+                    if let Some(peer_keepalive) = peer_keepalive_weak.upgrade() {
```

Comment:
Is there a situation where consensus_service would be running after the peer_keepalive has been dropped? I'm wondering if `upgrade().expect()` would be better than `if let Some`.

Reply:
yeah good call, that seems reasonable

Reply:
@mfaulk i guess the reason is, if someone calls `consensus_service.stop()` then this should get dropped, and that will cause the weak reference to return none. so if we panic when it becomes `none`, then the server will crash when we try to stop it

one way to deal with this might be, to try to change which of these objects are actually owning eachother. i think the only party that talks to the peer_keepalive variable may be, the `consensus_msgs_from_network` thread. so we could try to make that the exclusive owner, and just have it be an `Box` that gets moved into that thread

it seems to me that the main reason we have all these member variables of `consensus_service` is to support the `pub fn stop(&mut self)` thing, which stops all the threads without tearing down any of the objects. i'm not sure the `stop` thing is actually useful, it might be simpler to just stop the threads in the `drop` calls. That might reduce the amount of code overall, and in case like the peer_keepalive, it seems that we end up with mutexes only because we need exclusive access to stop the object, which doesn't seem like a good trade. `drop` has signature `drop(&mut self)` -- the language is ensuring that the call to the destructor is exclusive, so that's a good time to join things, IMO. that's just one opinion though, maybe there's a good use case for stopping without tearing down the objects.